### PR TITLE
Filter by previous players

### PR DESCRIPTION
**Pull request by MichaelThomasMPT**
Sometimes I've found that my friends aren't in the mood to learn new games and would rather play something they're familiar with. This PR adds an option to filter games by players who have played them previously, based on the user's logged plays (I log mine with the [BG Stats app](https://www.bgstatsapp.com/), for example). If you have a group of friends looking for something familiar, you can filter by all members in the group to see what games have been played by all.

I'm assuming most users would not want this functionality, especially not by default; for example, in case they are sharing their instance of the app publicly with others. Therefore I have hidden this functionality behind a "show_previous_players" toggle in the main config.json file; anybody who does not explicitly set this to true, including older config versions that do not set this field at all, will have this functionality disabled in both the front and back end. 

As the number of different players in the filter is variable from person to person, I couldn't find a good way to organise them in the refinement list so that all could be scrolled through; the best I could come up with was setting it as a searchable filter. I'm very open to any suggestions of how this could be improved! :)

### CODE REVIEW

**Comment by EmilStenstrom** [comment]
@MichaelThomasMPT Exciting! I currently don't log who I play with for each game, so I don't have any data to test this with. How would you use this in the UI?

**Comment by MichaelThomasMPT** [comment]
Feel free to try it out on my BGG instance I'm running:
https://boardgames.micker.xyz/

Or to test on your end, you might be able to point your config file to a different BGG who has plays logged - my username is Micker if you want to test with that.

**Comment by EmilStenstrom** [comment]
The code looks great. I'm not sure I will used it, as I only log plays, not who I played with. So ideally this would could be added as some kind of plugin, as I have lots of filter that I want to added, but haven't had the time to write yet.

But there is no plugin system, so lets do this anyways :)

The only thing I'm concerned with is the storage of the plays. I would like to add a filter where I can find games I haven't played yet. I think reusing the play data for that filter might be a good idea, but I think that requires the data to be added another way.

**Comment by MichaelThomasMPT** [comment]
Ah, I hadn't considered that use case. Let me have a think about that one and I'll take a look at it again tonight.

**Comment by MichaelThomasMPT** [comment]
After some further investigation into how InstantSearch works I'm not sure how to do an "exclude" at all in a refinementList. I've managed to get what you described above working via a toggleRefinement widget:

```
"refine_previousnonplayers": panel('Previous non-players')(instantsearch.widgets.toggleRefinement)(
      {
        container: '#facet-previous-nonplayers',
        attribute: 'previous_players',
        on: '-Michael Thomas',
        templates: {
          labelText: 'Not Played By Michael',
        },
      }
    ),
```

The problem with this is that it's hard-coded to just me, I haven't been able to find a way to populate this list dynamically with all the users who have plays recorded.

**Comment by MichaelThomasMPT** [comment]
I also found this very complicated example of implementing an exclusion refinementList, which would probably work, but not easily. 
https://codesandbox.io/s/4xvz9wmj00

**Comment by EmilStenstrom** [comment]
Thanks for the research! Since I don't add myself to plays. I just count the number of times I've played a certain game. But I think this might just be another aggregated number that I can add myself later. Does that make sense?

**Comment by MichaelThomasMPT** [comment]
Ah yeah - I've just stalked your BGG profile and I now see what you mean. :)

I think perhaps treating Plays and Players as two separate attributes is probably simpler - it means losing the link between which plays were associated with which players, but that doesn't seem to be that important for what this app is looking to do. I think perhaps having two attributes, one which lists the previous players, and another which counts the number of plays per game, would keep it simple and allow filtering on either. What are your thoughts?

I'll create a commit to add a play counter and see how that looks.

**Comment by MichaelThomasMPT** [comment]
Turns out numplays is a field which already exists on the response from the /collection endpoint, which makes it easy!

**Comment by MichaelThomasMPT** [comment]
You could perhaps use a range slider to restrict it down to a certain number of plays:
https://www.algolia.com/doc/api-reference/widgets/range-slider/js/

**Comment by EmilStenstrom** [comment]
Definitely! Thanks for adding more filters for me :)

**Comment by MichaelThomasMPT** [comment]
No worries! Would you like me to add it as a filter in the UI too? If so, which do you think would work better - a range slider, or a refinementList like the others, but with play number categories (maybe every 0, 1-10, 11-20, 21-30 etc.)?

**Comment by MichaelThomasMPT** [comment]
Turns out the range slider won't meet your needs - looks like you can't have a range of 0 to 0, the smallest is 0 to 1, which is inclusive. A [rangeInput widget](https://www.algolia.com/doc/api-reference/widgets/range-input/js/) would work, but it looks a little uglier.

**Comment by EmilStenstrom** [comment]
Hi! I think a numericMenu would work great.

**Comment by MichaelThomasMPT** [comment]
Hey @EmilStenstrom - I've added a numericMenu to filter on total plays for each game. 
I've pushed those changes to [my MyBGG instance](https://boardgames.micker.xyz) if you want to check it out there.

File: app.js [hunk 1]
```diff
@@ -180,6 +180,30 @@ function get_widgets(SETTINGS) {
...
+      {
+        container: '#facet-numplays',
+        attribute: 'numplays',
+        items: [
+          { label: 'Any' }, 
+          { label: '0', end: 0 },
+          { label: '1', start: 1, end: 1 },
+          { label: '2-9', start: 2, end: 9 },
+          { label: '10-19', start: 10, end: 19 },
+          { label: '20-39', start: 20, end: 29 },
```
Comment (EmilStenstrom) [inline app.js]: Typo: should be 29, not 39.

**Comment by EmilStenstrom** [comment]
@MichaelThomasMPT Found a typo, then I think this is done! :) Big thanks for all your work put into this.

File: app.js (hunk 1 above)
Comment (MichaelThomasMPT) [reply app.js]: Oops, good catch. I've gotta get better at proof-reading before hitting commit!

Comment (EmilStenstrom) [reply app.js]: That's why we have code review! :)

**Comment by MichaelThomasMPT** [comment]
No worries! Thank for taking the time to review & provide great feedback. :)

**Comment by EmilStenstrom** [comment]
FYI: I made some simplifications after playing with a code a bit more:

* I removed the two new settings again. Instead the filters for previous players and total plays show up automatically when there is data. The idea is that more users will get the new filters that way, without having to fiddle with settings. I've seen many users struggle with config.json and getting it commited to git, this saves them from that
* I made sure the replica indexes get your configured algolia index name as prefix. Is just seamed nicer to have everything prefixed with the same key.

**Comment by MichaelThomasMPT** [comment]
Looks good, thanks for letting me know!

**Pull request merged**